main/src/ca/uwaterloo/flix/api/lsp/provider/HighlightProvider.scala

```diff
@@ -15,195 +15,273 @@
 import org.json4s.JsonDSL.*
 
+import scala.annotation.tailrec
+
 object HighlightProvider {
 
-  def processHighlight(uri: String, pos: Position)(implicit index: Index, root: Root): JObject = {
-    index.query(uri, pos) match {
+  def processHighlight(uri: String, pos: Position)(implicit root: Root): JObject = {
+    val stackConsumer = StackConsumer()
+    Visitor.visitRoot(root, stackConsumer, Visitor.InsideAcceptor(uri, pos))
+
+    stackConsumer.getStack.headOption match {
       case None => mkNotFound(uri, pos)
+      case Some(x) => highlightAny(uri, pos, x)
+    }
 
-      case Some(entity) => entity match {
-        case Entity.Case(caze) => highlightCase(uri, caze.sym)
+  }
 
-        case Entity.StructField(field) => highlightStructField(uri, field.sym)
+  @tailrec
+  private def getTypeSymOccur(tpe: Type): Option[(Symbol, SourceLocation)] = tpe match {
+    case Type.Cst(TypeConstructor.Enum(sym, _), loc) => Some(sym, loc)
+    case Type.Cst(TypeConstructor.Struct(sym, _), loc) => Some(sym, loc)
+    case Type.Cst(TypeConstructor.Effect(sym), loc) => Some(sym, loc)
+    case Type.Apply(tpe1, _, _) => getTypeSymOccur(tpe1)
+    case _ => None
+  }
 
-        case Entity.Def(defn) => highlightDef(uri, defn.sym)
+  private def highlightAny(uri: String, pos: Position, x: AnyRef)(implicit root: Root): JObject = x match {
+    // Variables
+    case Binder(sym, _) => highlightVarSym(uri, sym, sym.loc)
+    case TypedAst.Expr.Var(varSym, _, loc) => highlightVarSym(uri, varSym, loc)
+    // Enums & Cases
+    case TypedAst.Enum(_, _, _, sym, _, _, _, _) => highlightEnumSym(uri, sym, sym.loc)
+    case TypedAst.Case(sym, _, _, _) => highlightCaseSym(uri, sym, sym.loc)
+    case SymUse.CaseSymUse(sym, loc) => highlightCaseSym(uri, sym, loc)
+    // Defs
+    case TypedAst.Def(sym, _, _, _) => highlightDefnSym(uri, sym, sym.loc)
+    case SymUse.DefSymUse(sym, loc) => highlightDefnSym(uri, sym, loc)
+    // Structs
+    case TypedAst.Struct(_, _, _, sym, _, _, _, _) => highlightStructSym(uri, sym, sym.loc)
+    case TypedAst.StructField(sym, _, _) => highlightStructFieldSym(uri, sym, sym.loc)
+    case SymUse.StructFieldSymUse(sym, loc) => highlightStructFieldSym(uri, sym, loc)
+    // Effects
+    case TypedAst.Effect(_, _, _, sym, _, _) => highlightEffectSym(uri, sym, sym.loc)
+    case SymUse.EffectSymUse(sym, loc) => highlightEffectSym(uri, sym, loc)
+    // Traits
+    case TypedAst.Trait(_, _, _, sym, _, _, _, _, _, _) => highlightTraitSym(uri, sym, sym.loc)
+    case SymUse.TraitSymUse(sym, loc) => highlightTraitSym(uri, sym, loc)
+    case TraitConstraint.Head(sym, loc) => highlightTraitSym(uri, sym, loc)
+    case Ast.Derivation(sym, loc) => highlightTraitSym(uri, sym, loc)
+    // Signatures
+    case TypedAst.Sig(sym, _, _, _) => highlightSigSym(uri, sym, sym.loc)
+    case SymUse.SigSymUse(sym, loc) => highlightSigSym(uri, sym, loc)
+    // Symbols in types
+    case tpe: Type => getTypeSymOccur(tpe) match {
```

Comment:
Extend visitor with:

visitTypeCst
visitTypeVar
visitTypeApply
... type alias and assoc type ...

Reply:
Yes. This has been done in #9115